### ISSUE TITLE
Any HTTP service will block all HTTPS/TCP traffic on the same port

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#16458]
(Assumes ALLOW_ANY mode)

Normally, traffic to external HTTPS services works due to the ALLOW_ANY mode changes we added. However, when an http service is added on port 443 (or any port, but generally this happens on 443), this breaks. This is because we add a new 0.0.0.0_443 listener, which will match everything and direct to routes. This will fail.

To reproduce:
```
curl https://www.google.com # works
```
then
```
cat <<EOF | kaf -
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: https-breaker
spec:
  hosts:
  - wikipedia.org
  location: MESH_EXTERNAL
  ports:
  - number: 443
    name: http
    protocol: HTTP
  resolution: DNS
EOF
```
now

```
$ curl https://www.google.com
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol.
```

One possible idea is to do similar to protocol sniffing and detect TLS connection and direct to passthrough cluster. This may be possible with tls inspector.

Related issues:
* https://github.com/istio/istio/issues/16450
* https://github.com/istio/istio/issues/14520
* https://github.com/istio/istio/issues/16196
* https://github.com/istio/istio/issues/16331
* https://github.com/istio/istio/issues/13914

**Comment by howardjohn** [comment]
@yxue think this is something that could get tied in with your protocol sniffing work?

**Comment by yxue** [comment]
Technically protocol sniffing can resolve these issue but it requires tons of efforts. 

To resolve these problem, we need to sniff the traffic using tls_inspector. The tough part is that we need to generate multiple clusters and multiple routes. Currently, the protocol sniffing only generates multiple filter chains without touching clusters and routes. For this issue, we need to generate two upstream clusters, one with tls_context and another one without. Also, we need to generate two routes one used for https and one used for http. 

Protocol sniffing for h2 is similar. That's why protocol sniffing for h2 is not support yet.

**Comment by howardjohn** [comment]
@yxue for these we just need to send to passthrough cluster though, since this is just for external ALLOW_ANY traffic. If a user has an explicitly defined https service it should still work because it will have a filter chain match on the SNI name, right?

**Comment by vadimeisenbergibm** [comment]
@howardjohn First note that defining HTTP for 443 port is usually a mistake, the port and protocol in your example of wikipedia.org should be HTTPS. Second, I think the same issue can happen inside the mesh, if for some port inside the mesh HTTP and non-HTTP protocols will be used for the same port.

**Comment by howardjohn** [comment]
@vadimeisenbergibm my understanding is if you try to talk to an https service in-mesh it should work, because we will generate a listener with the SNI match that will match before it matches the HTTP routes (Which is where the issue arises). Untested though

**Comment by vadimeisenbergibm** [comment]
@howardjohn Agreed.

**Comment by Dev25** [comment]
@howardjohn  @vadimeisenbergibm 

I've also ran into this last week when trying to do egress TLS origination + ALLOW_ANY so we can slowly integrate external services with istio capabilities, FYI regarding the HTTP on port 443 the current docs explicity mention to do that as part of tls origination.

I've not yet tested what happens when the SE's 443 port is set to HTTPS whilst doing tls origination.

https://istio.io/docs/tasks/traffic-management/egress/egress-tls-origination/#tls-origination-for-egress-traffic
> As you can see, the VirtualService redirects HTTP requests on port 80 to port 443 where the corresponding DestinationRule then performs the TLS origination. Notice that unlike the ServiceEntry in the previous section, this time the protocol on port 443 is HTTP, instead of HTTPS. This is because clients will only send HTTP requests and Istio will upgrade the connection to HTTPS.

**Comment by Dev25** [comment]
Update: Setting the ServiceEntry port to HTTPS works fine with TLS Origination, so this is also a documentation issue.

**Comment by vadimeisenbergibm** [comment]
@Dev25 You mean setting the 443 port to be HTTPS works fine with TLS origination? So the documentation must be fixed.

**Comment by Dev25** [comment]
@vadimeisenbergibm Yep setting ServiceEntry port 443 to HTTPS works just fine with doing TLS origination.

That being said i'm not sure if this is intended behaviour but trying to connect to port 443 using HTTPS directly (when configured for tls origination) will result in connection failures.

```
Python/Requests:
Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:852)'),)

wget:
ssl_client: <domain>: handshake failed: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
wget: error getting response: Connection reset by peer
```

Access logs do get generated by istio-proxy for these failed attempts
```json
{
  "authority": "-",
  "bytes_received": "439",
  "bytes_sent": "890",
  "downstream_local_address": "<redacted ip>:443",
  "downstream_remote_address": "10.240.2.70:34604",
  "duration": "20",
  "istio_policy_status": "-",
  "method": "-",
  "path": "-",
  "protocol": "-",
  "request_id": "-",
  "requested_server_name": "api.mydomain.com",
  "response_code": "0",
  "response_flags": "-",
  "start_time": "2019-08-28T13:53:44.234Z",
  "upstream_cluster": "outbound|443||api.mydomain.com",
  "upstream_host": "<redacted ip>:443",
  "upstream_local_address": "10.240.2.70:34610",
  "upstream_service_time": "-",
  "upstream_transport_failure_reason": "-",
  "user_agent": "-",
  "x_forwarded_for": "-"
}
```

**Comment by vadimeisenbergibm** [comment]
@Dev25
> trying to connect to port 443 using HTTPS directly (when configured for tls origination) will result in connection failures.

this is totally by design - the gateway expects HTTP traffic, so sending it HTTPS traffic will not work.

**Comment by howardjohn** [comment]
One alternative solution -- reject http Services on 443. Not sure if that is a worse evil than the bug though...

Or treat `http` on 443 as https or tcp (either one works). Both of these are pretty hacky

**Comment by howardjohn** [comment]
Spoke with @costinm and @andraxylia , seems rejecting http Services on port 443 is likely the best move here. Specifically this would involve rejecting ServiceEntries with this pattern in Galley validation, then rejecting Services (or ServiceEntries that bypass validation, I suppose) internally in pilot (and definitely increment `pilot_total_rejected_configs`! or some metric)

**Comment by rshriram** [comment]
we have fixed these conflicts with the http inspector.. right @yxue ?

**Comment by howardjohn** [comment]
This is not fixed

**Comment by howardjohn** [comment]
We will create a listener with HCM only. We need to detect HTTPS traffic and not send them through HCM

**Comment by viktorpeacock** [comment]
Hello.

This affects us as we use TLS origination using the provided documentation, i.e. having HTTP on port 443 and then using destination route to originate TLS.

This has stopped working recently and I can only assume this was caused by the ISTIO upgrade by the infrastructure team. 

Dev25 suggested that changing port to HTTPS works, but it didn't for us. Can you please provide some clarify on whether this will change or how it can be fixed as we cannot talk to the external service whilst using virtual services for the retries.

Thank you.

**Comment by vadimeisenbergibm** [comment]
@viktorpeacock let me check this. So the problem started to appear after upgrade to Istio 1.3.0?

**Comment by vadimeisenbergibm** [comment]
@viktorpeacock For me [specifying the port as HTTPS or TLS](https://istio.io/docs/tasks/traffic-management/egress/egress-tls-origination/#tls-origination-for-egress-traffic) works in Istio 1.3.0. I will update the configuration.

**Comment by y0zg** [comment]
> @viktorpeacock For me [specifying the port as HTTPS or TLS](https://istio.io/docs/tasks/traffic-management/egress/egress-tls-origination/#tls-origination-for-egress-traffic) works in Istio 1.3.0. I will update the configuration.

@vadimeisenbergibm could it help if HTTP/HTTPS are on the same port?

**Comment by goruha** [comment]
We had the problem with https external service call on a default installation of Istio with helm chart version `1.3.3`
The reason was in enabled by default `enableProtocolSniffingForOutbound`
https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/charts/pilot/values.yaml#L16

I belive for `ALLOW_ANY` mode `enableProtocolSniffingForOutbound` should be `false` like `enableProtocolSniffingForInbound` is.

**Comment by howardjohn** [comment]
@goruha I don't think the two features are related at all, why do you think it broke? Can you provide more details about your setup and the config dump? Probably in another issue since it doesn't seem related

**Comment by goruha** [comment]
@howardjohn
Platform AWS

Schema looks like that 
ELB (https termination) -> Service -> Istio Gateway -> VirtualService -> Pod

Ingress Gateway service

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    external-dns.alpha.kubernetes.io/hostname: '*.example.com'
    external-dns.alpha.kubernetes.io/ttl: "300"
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: arn:aws:acm:eu-west-2:XXXXXX:certificate/XXXXX
    service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: ELBSecurityPolicy-TLS-1-2-2017-01
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: https
  creationTimestamp: "2019-09-20T10:19:40Z"
  labels:
    app: istio-ingressgateway
    chart: gateways
    heritage: Tiller
    istio: ingressgateway
    release: istio
  name: istio-ingressgateway
  namespace: istio-system
  resourceVersion: "51793146"
  selfLink: /api/v1/namespaces/istio-system/services/istio-ingressgateway
  uid: 280f42d0-db90-11e9-be4f-06c7a02953b4
spec:
  clusterIP: 100.68.205.205
  externalTrafficPolicy: Cluster
  ports:
  - name: status-port
    nodePort: 32570
    port: 15020
    protocol: TCP
    targetPort: 15020
  - name: http2
    nodePort: 31380
    port: 80
    protocol: TCP
    targetPort: 80
  - name: https
    nodePort: 31390
    port: 443
    protocol: TCP
    targetPort: 443
  - name: tcp
    nodePort: 31400
    port: 31400
    protocol: TCP
    targetPort: 31400
  - name: https-kiali
    nodePort: 30073
    port: 15029
    protocol: TCP
    targetPort: 15029
  - name: https-prometheus
    nodePort: 30667
    port: 15030
    protocol: TCP
    targetPort: 15030
  - name: https-grafana
    nodePort: 31968
    port: 15031
    protocol: TCP
    targetPort: 15031
  - name: https-tracing
    nodePort: 31662
    port: 15032
    protocol: TCP
    targetPort: 15032
  - name: tls
    nodePort: 30441
    port: 15443
    protocol: TCP
    targetPort: 15443
  selector:
    app: istio-ingressgateway
    istio: ingressgateway
    release: istio
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - hostname: XXXXXX.eu-west-2.elb.amazonaws.com
```

Ingress Gateway

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  creationTimestamp: "2019-10-17T17:37:47Z"
  generation: 3
  labels:
    app: raw
    chart: raw-0.1.0
    heritage: Tiller
    release: istio-additional
  name: istio-ingressgateway
  namespace: istio-system
  resourceVersion: "51792792"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/istio-system/gateways/istio-ingressgateway
  uid: d566be97-f104-11e9-adbc-0a8a771e0a30
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: http
      number: 80
      protocol: HTTP
    tls:
      httpsRedirect: true
  - hosts:
    - '*'
    port:
      name: https
      number: 443
      protocol: HTTP
```

VirtualService

```
kind: VirtualService
apiVersion: networking.istio.io/v1alpha3
metadata:
  name: web-api-gateway-blue-default
  namespace: prod
  resourceVersion: '51792119'
  generation: 1
  creationTimestamp: '2019-10-17T17:53:40Z'
  labels:
    app: web-api-gateway
    chart: monochart-0.18.2
    heritage: Tiller
    release: prod-web-api-gateway
spec:
  hosts:
    - web-api-gateway.example.com
  gateways:
    - istio-system/istio-ingressgateway
  http:
    - match:
        - uri:
            prefix: /
      name: default
      route:
        - destination:
            host: web-api-gateway-blue
            port:
              number: 80
          weight: 100
  tcp: ~
  tls: ~
  exportTo: ~
```

Everything else is default chart values you can find [here](https://github.com/cloudposse/helmfiles/blob/istio-egress-issues/releases/istio.yaml)

When I have [`enableProtocolSniffingForOutbound: false`](https://github.com/cloudposse/helmfiles/blob/istio-egress-issues/releases/istio.yaml#L219)
everything works.
If it is `true` we got 
```
bash-4.4# curl https://google.com
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
```

That probably different issue, but related.

Why `enableProtocolSniffingForOutbound` should be `true` by default?

**Comment by goruha** [comment]
@howardjohn another workaround to make it work is to create `ServiceEntry` for `google.com`, but in case of `ALLOW_ANY` such entity can not be required for correct behaviour

**Comment by howardjohn** [comment]
@goruha can you capture a config dump and create another issue and assign/mention @yxue ? instructions https://github.com/istio/istio/wiki/Troubleshooting-Istio#collecting-information-2

**Comment by jonmoter** [comment]
We are seeing similar behavior.  We have a Kubernetes `Service` deployed in our cluster that has a port listening on port 443, and the port is named `grpc`.  When that Service is present, any pod that has the istio sidecar injected is unable to make outbound requests.

```
$ kubectl -n mesh-enabled exec -it ${pod_name} -- curl -I https://www.google.com
curl: (35) error:1400410B:SSL routines:CONNECT_CR_SRVR_HELLO:wrong version number
command terminated with exit code 35
```

But if I rename the Service's port from `grpc` to `https-grpc`, then it works, and the same curl command above succeeds.

So it's not just naming the port `http` that triggers this issue.

**Comment by howardjohn** [comment]
Quick update - short term solution to reject http ports on port 80 is shipped in 1.4 (on by default) and in 1.3.5 (off by default). Long term fix still pending design/implmentation

**Comment by Nuru** [comment]
@howardjohn wrote
> Quick update - short term solution to reject http ports on port 80 is shipped in 1.4 (on by default) and in 1.3.5 (off by default). Long term fix still pending design/implementation

Would you please elaborate? 
- How does rejecting HTTP ports on port 80 help anything?
- Are we any closer to a long-term fix? 
- The link you referenced earlier ( https://preliminary.istio.io/news/2019/announcing-1.4/upgrade-notes/#traffic-management ) is broken. What does it say? 

Also
- How does this issue relate to #18110?
- How does this issue relate to #20703?

Is it currently possible to enable tracing/telemetry of some external HTTPS services while still having a default policy of allowing them all? This and the other referenced bugs had the effect that, unlike with HTTP services, adding controlled access to a single external HTTPS service effectively made all the HTTPS services `REGISTRY_ONLY` even if `ALLOW_ANY` was set. If this has been fixed, please say where and how.

**Comment by howardjohn** [comment]
sorry I meant port 443not 80. Yes you can add an external service for https, just name the port https not http

**Comment by camattin** [comment]
I need to +1 this issue.

Specifically what we are seeing in our environment is whenever there is a service of type http defined anywhere in the mesh then you cannot access anything outside the mesh on that same port when it is TLS.

As an example our outbound proxies listen on port 8080 which many services in the mesh like to listen on as well. And those services are correctly named as http-<whatever>. Now when a pod tries to utilize outboundproxy.company.com:8080 (via http(s)_proxy env vars already set) it only works for  reaching http sites (and the Server: header in the traffic confirms that some service's envoy intercepted it, but it still worked). Accessing an https site fails.

I'm using the proxy example here, but this means that any user in a cluster can define a service of type http on any port and that will break anyone from accessing external services on that same port if they are TLS.

Having to manage ServiceEntries for this **is not a manageable solution**. Even more so because "resolution: DNS" does not appear to be working either -- we're having to add the specific target IPs into the addresses instead of simply specifying the hostname in the hosts: section.

We've also experimented with includeIPRanges on the helm install and while that sets the range on the initContainer's istio-iptables-go command it still does not make the change functional. However, if we set it with a includeOutboundIPRanges annotation it /does/ work. Because the --set option is seemingly not functional, having to dictate that every deployment must include this annotation is also not ideal.

Note: We are using istio-cni and have this behavior with with 1.4.4 and 1.4.5.

**Comment by howardjohn** [comment]
There is not any mitigation that I know of right now other than:
* only sending TLS traffic on port 443 (I agree this is not good - just the way things are right now)
* creating service entries

**Comment by hzxuzhonghu** [comment]
Can we add an additional filterchain with filterchainmatch.transport_protocol = tls to match https/tls?

**Comment by howardjohn** [comment]
Yes I think that's the proper fix, seems like it should work?

**Comment by hzxuzhonghu** [comment]
For http/tcp protocol, if the destination addresses are mesh external, then we can add a filter chain to passthrough?

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/17169 should fix this, but I will leave this open to track adding additional testing to ensure we have covered all cases

**Comment by zhaohuabing** [comment]
Is this the ideal listener and filter structure?

![istio-listener-structure](https://user-images.githubusercontent.com/1321762/77710144-3f2a2f80-7008-11ea-9a91-109300be4116.png)

<details>
https://edotor.net/?engine=dot?engine=dot#digraph%20%7B%0A%20%20%20%20%23%20Application%20traffic%20is%20capture%20by%20iptables%2C%20redirected%20to%20port%2015001%0A%20%20%20%20app%20-%3E%20virtualOutbound%0A%0A%20%20%20%20%23%20We%20create%20a%20listener%20on%200.0.0.0%20to%20catch%20all%20traffic%20on%20a%20specific%20port-%20matching%20is%20done%20in%20routes(HTTP)%20or%20destination%20IP(TCP)%0A%20%20%20%20virtualOutbound%20-%3E%20wildcardListener%20%5Blabel%3D%22Match%20Port%22%2Ccolor%3Dgreen%5D%0A%0A%20%20%20%20%20%23%20For%20HTTP%20services%2C%20we%20will%20construct%20a%20wildcard%20match%20-%20routing%20done%20in%20RDS%0A%20%20%20%20wildcardListener%20-%3E%20routePort%20%5Blabel%3D%22Sniffed%20HTTP%22%5D%0A%20%20%20%20routePort%20-%3E%20routeMatchQ%20%5Blabel%3D%22Domain%20Match%22%2Ccolor%3Dgreen%5D%0A%20%20%20%20routePort%20-%3E%20fallbackRouteQ%20%5Blabel%3D%22No%20Match%22%2Ccolor%3Dred%5D%0A%0A%20%20%20%20%23%20If%20domain%20matches%2C%20we%20then%20check%20routes%20match%0A%20%20%20%20routeMatchQ%20-%3E%20noRoute%20%5Blabel%3D%22No%20Match%22%2Ccolor%3Dred%5D%0A%20%20%20%20routeMatchQ%20-%3E%20upstreamCluster%20%5Blabel%3D%22Match%22%2Ccolor%3Dgreen%5D%0A%0A%20%20%20%20%20%23%20We%20fallback%20to%20passthrough%20or%20blackhole%2C%20depending%20on%20config%0A%20%20%20%20fallbackRouteQ%20-%3E%20passthrough%20%5Blabel%3D%22ALLOW_ANY%22%5D%0A%20%20%20%20fallbackRouteQ%20-%3E%20blackholeRoute%20%5Blabel%3D%22REGISTRY_ONLY%22%5D%0A%20%0A%20%20%20%20%23%20For%20TCP%20services%2C%20Routing%20is%20done%20with%20%20IP%20destination%20of%20filterChainMatch%0A%20%20%20%20wildcardListener%20-%3E%20destionationIPMatchQ%20%5Blabel%3D%22Sniffed%20TCP%22%5D%0A%20%20%20%20destionationIPMatchQ%20-%3E%20tcpProxy%5Blabel%3D%22Match%22%5D%20%20%20%20tcpProxy%20-%3E%20upstreamCluster%20%0A%20%20%20%20destionationIPMatchQ%20-%3E%20fallbackQ%20%5Blabel%3D%22No%20Match%22%5D%0A%0A%20%20%20%20%20%23%20We%20fallback%20to%20passthrough%20or%20blackhole%2C%20depending%20on%20config%0A%20%20%20%20fallbackQ%20-%3E%20passthrough%20%5Blabel%3D%22ALLOW_ANY%22%5D%0A%20%20%20%20fallbackQ%20-%3E%20blackhole%20%5Blabel%3D%22REGISTRY_ONLY%22%5D%0A%20%20%0A%20%20%20%20%23%23%23%20Node%20Styling%0A%20%20%20%20app%20%5Blabel%3D%22Application%20(Outbound)%22%5D%0A%20%20%20%20virtualOutbound%20%5Blabel%3D%22virtualOutbound%20(15001)%22%5D%0A%0A%20%20%20%20routeMatchQ%20%5Bshape%3D%22diamond%22%2C%20label%3D%22Route%20Match%3F%22%5D%0A%20%20%20%20fallbackRouteQ%20%5Bshape%3D%22diamond%22%2C%20label%3D%22OutboundTrafficPolicy%3F%22%5D%0A%20%20%20%20blackholeRoute%20%5Blabel%3D%22502%20Response%22%2Ccolor%3D%22red%22%5D%0A%0A%20%20%20%20fallbackQ%20%5Bshape%3D%22diamond%22%2C%20label%3D%22OutboundTrafficPolicy%3F%22%5D%0A%20%20%20%20passthrough%20%5Blabel%3D%22Passthrough%20Cluster%22%2Ccolor%3D%22green%22%5D%0A%20%20%20%20blackhole%20%5Blabel%3D%22Blackhole%20Cluster%22%2Ccolor%3D%22red%22%5D%0A%0A%20%20%20%20tcpProxy%20%5Blabel%3D%22TCP%20proxy%22%5D%0A%0A%20%20%20%20routePort%20%5Blabel%3D%22HTTP%20Route%20for%20%3CPORT%3E%22%5D%0A%0A%20%0A%20%20%20%20wildcardListener%20%5Blabel%3D%220.0.0.0_%3CPORT%3E%20listener%22%5D%0A%0A%20%20%20%20noRoute%20%5Blabel%3D%22404%20Response%22%2C%20color%3D%22red%22%5D%0A%20%20%20%20upstreamCluster%20%5Blabel%3D%22Upstream%20Cluster%22%2C%20color%3D%22green%22%5D%0A%0A%20%20%20%20destionationIPMatchQ%20%5Bshape%3D%22diamond%22%2C%20label%3D%22Destination%20IP%20Match%3F%22%5D%0A%7D%0A
</details>

**Comment by zhaohuabing** [comment]
OK, I think we can't do this unless protocol sniffing is always enabled.

**Comment by hustshawn** [comment]
> Its fine to say that you can't have http on port 443, but if that is the case we should explicitly block it -- otherwise one service in some random namespace can destroy a whole cluster

How do we achieve it?

**Comment by howardjohn** [comment]
In 1.4+ http ports on 443 are blocked. In 1.6 this will not be needed as this bug has properly been fixed

**Comment by ramaraochavali** [comment]
> In 1.6 this will not be needed as this bug has properly been fixed

Is it because by default we enable protocol sniffing for outbound?

**Comment by howardjohn** [comment]
Yes and we added a sniffing part to the wildcard listeners, if tcp is
matched we just pass through

**Comment by ramaraochavali** [comment]
Should we change the default of `PILOT_BLOCK_HTTP_ON_443` this flag then in 1.6?

**Comment by howardjohn** [comment]
I was planning to remove that today once I make sure we get proper testing
of it. We sort of are testing it today but not fully

**Comment by ramaraochavali** [comment]
Great.

**Comment by camattin** [comment]
I'm a little confused on today's updates... is this going to fix the issue of outbound TLS connections where it happens to be on the same port as _any_ http-* named service in the mesh? Port number does not matter here...

**Comment by howardjohn** [comment]
@camattin yes, this should be fixed now.

**Comment by camattin** [comment]
Excellent, is this in early enough for 1.5.2?

**Comment by howardjohn** [comment]
It is not currently slated for a 1.5.x release. I am working right now on improving the test coverage of this change/issue in general. Once that is done its possible we can backport the change behind a feature flag since its a pretty major pain point for users

**Comment by hugree** [comment]
Is there a working workaound for this or not? I've seen mixed responses.
I'm trying to use TLS origination in order to achieve a working timeout on a request to external service via HTTPS 

I use the following config:
```apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: httpbin-ext
spec:
  hosts:
    - httpbin.org
  ports:
    - number: 80
      name: http-httpbin
      protocol: HTTP
    - number: 443
      name: https-port-for-tls-origination
      protocol: HTTPS
  resolution: DNS
  location: MESH_EXTERNAL
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: httpbin-ext
spec:
  hosts:
    - httpbin.org
  http:
    - match:
        - port: 80
      route:
        - destination:
            host: httpbin.org
            port:
              number: 443
      timeout: 2.500s
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: httpbin-ext
spec:
  host: httpbin.org
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    portLevelSettings:
    - port:
        number: 443
      tls:
        mode: SIMPLE
```
In this case
`time curl -o /dev/null -s -w "%{http_code}\n" http://httpbin.org/delay/3`
gives me a proper behavior with timeout protection
```
504

real	0m2.519s
user	0m0.004s
sys	0m0.004s
```
But going for HTTPS ends up with an error
```
curl https://httpbin.org/delay/3
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

**Comment by KinglyWayne** [comment]
This bug seems to happen again with istio version 1.6.7
One http service with service port 443 would still break all cluster external https access.

**Comment by howardjohn** [comment]
Yes we recently found this is still broken in the (fairly common) case of setting ALPN in the request on an https call. Being tracked in https://github.com/istio/istio/issues/26079. The fix may require some work in envoy which will make it challenging

**Comment by shalk** [comment]
I meet a similar problem at 1.5.0 

**headless service's port name without  protocol-suffix**  will block external https traffic on the same port;

It looks like headdless service without protocol-suffix will create lds  ` 0.0.0.0_port` . this lds will not properly handle externl https.

I can reproduce it.

headless service yaml as follow:
```
apiVersion: v1
kind: Service
metadata:
  name: mysvc
  namespace: service-mesh-test
spec:
  clusterIP: None
  ports:
  - name: mysvc
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    ocean-app: mysvc
  type: ClusterIP

```

```
k -n service-mesh-test get svc mysvc
NAME    TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
mysvc   ClusterIP   None         <none>        443/TCP   31m

```

```
kn exec -ti sleep-6c59fc9c89-x5szc  -- curl https://www.baidu.com                         
Defaulting container name to sleep.
Use 'kubectl describe pod/sleep-6c59fc9c89-x5szc -n service-mesh-test' to see all of the containers in this pod.
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
command terminated with exit code 35
```

the LDS is as follow,   If I changed  headdless service's port name with protocol-suffix (eg. tcp,http,https) ， LDS will disappear and  external https will passthrough

```
istioctl proxy-config listener sleep-6c59fc9c89-x5szc --port 443 --address 0.0.0.0 -o json
[
    {
        "name": "0.0.0.0_443",
        "address": {
            "socketAddress": {
                "address": "0.0.0.0",
                "portValue": 443
            }
        },
        "filterChains": [
            {
                "filterChainMatch": {
                    "prefixRanges": [
                        {
                            "addressPrefix": "10.46.12.92",
                            "prefixLen": 32
                        }
                    ]
                },
                "filters": [
                    {
                        "name": "envoy.filters.network.wasm",
                        "typedConfig": {
                            "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                            "typeUrl": "type.googleapis.com/envoy.config.filter.network.wasm.v2.Wasm",
                            "value": {
                                "config": {
                                    "configuration": "{\n  \"debug\": \"false\",\n  \"stat_prefix\": \"istio\",\n}\n",
                                    "root_id": "stats_outbound",
                                    "vm_config": {
                                        "code": {
                                            "local": {
                                                "inline_string": "envoy.wasm.stats"
                                            }
                                        },
                                        "runtime": "envoy.wasm.runtime.null",
                                        "vm_id": "stats_outbound"
                                    }
                                }
                            }
                        }
                    },
                    {
                        "name": "envoy.tcp_proxy",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
                            "statPrefix": "BlackHoleCluster",
                            "cluster": "BlackHoleCluster"
                        }
                    }
                ]
            },
            {
                "filters": [
                    {
                        "name": "envoy.filters.network.wasm",
                        "typedConfig": {
                            "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                            "typeUrl": "type.googleapis.com/envoy.config.filter.network.wasm.v2.Wasm",
                            "value": {
                                "config": {
                                    "configuration": "{\n  \"debug\": \"false\",\n  \"stat_prefix\": \"istio\",\n}\n",
                                    "root_id": "stats_outbound",
                                    "vm_config": {
                                        "code": {
                                            "local": {
                                                "inline_string": "envoy.wasm.stats"
                                            }
                                        },
                                        "runtime": "envoy.wasm.runtime.null",
                                        "vm_id": "stats_outbound"
                                    }
                                }
                            }
                        }
                    },
                    {
                        "name": "envoy.tcp_proxy",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
                            "statPrefix": "outbound|443||mysvc.service-mesh-test.svc.cluster.local",
                            "cluster": "outbound|443||mysvc.service-mesh-test.svc.cluster.local"
                        }
                    }
                ]
            },
            {
                "filterChainMatch": {
                    "applicationProtocols": [
                        "http/1.0",
                        "http/1.1",
                        "h2c"
                    ]
                },
                "filters": [
                    {
                        "name": "envoy.http_connection_manager",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
                            "statPrefix": "outbound_0.0.0.0_443",
                            "rds": {
                                "configSource": {
                                    "ads": {}
                                },
                                "routeConfigName": "443"
                            },
                            "httpFilters": [
                                {
                                    "name": "envoy.filters.http.wasm",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                                        "typeUrl": "type.googleapis.com/envoy.config.filter.http.wasm.v2.Wasm",
                                        "value": {
                                            "config": {
                                                "configuration": "envoy.wasm.metadata_exchange",
                                                "vm_config": {
                                                    "code": {
                                                        "local": {
                                                            "inline_string": "envoy.wasm.metadata_exchange"
                                                        }
                                                    },
                                                    "runtime": "envoy.wasm.runtime.null"
                                                }
                                            }
                                        }
                                    }
                                },
                                {
                                    "name": "istio.alpn",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/istio.envoy.config.filter.http.alpn.v2alpha1.FilterConfig",
                                        "alpnOverride": [
                                            {
                                                "alpnOverride": [
                                                    "istio-http/1.0",
                                                    "istio"
                                                ]
                                            },
                                            {
                                                "upstreamProtocol": "HTTP11",
                                                "alpnOverride": [
                                                    "istio-http/1.1",
                                                    "istio"
                                                ]
                                            },
                                            {
                                                "upstreamProtocol": "HTTP2",
                                                "alpnOverride": [
                                                    "istio-h2",
                                                    "istio"
                                                ]
                                            }
                                        ]
                                    }
                                },
                                {
                                    "name": "envoy.cors"
                                },
                                {
                                    "name": "envoy.fault"
                                },
                                {
                                    "name": "envoy.filters.http.wasm",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                                        "typeUrl": "type.googleapis.com/envoy.config.filter.http.wasm.v2.Wasm",
                                        "value": {
                                            "config": {
                                                "configuration": "{\n  \"debug\": \"false\",\n  \"stat_prefix\": \"istio\",\n}\n",
                                                "root_id": "stats_outbound",
                                                "vm_config": {
                                                    "code": {
                                                        "local": {
                                                            "inline_string": "envoy.wasm.stats"
                                                        }
                                                    },
                                                    "runtime": "envoy.wasm.runtime.null",
                                                    "vm_id": "stats_outbound"
                                                }
                                            }
                                        }
                                    }
                                },
                                {
                                    "name": "envoy.router"
                                }
                            ],
                            "tracing": {
                                "clientSampling": {
                                    "value": 100
                                },
                                "randomSampling": {
                                    "value": 1
                                },
                                "overallSampling": {
                                    "value": 100
                                }
                            },
                            "streamIdleTimeout": "0s",
                            "useRemoteAddress": false,
                            "generateRequestId": true,
                            "upgradeConfigs": [
                                {
                                    "upgradeType": "websocket"
                                }
                            ],
                            "normalizePath": true
                        }
                    }
                ]
            }
        ],
        "deprecatedV1": {
            "bindToPort": false
        },
        "listenerFilters": [
            {
                "name": "envoy.listener.tls_inspector"
            },
            {
                "name": "envoy.listener.http_inspector"
            }
        ],
        "listenerFiltersTimeout": "0.100s",
        "continueOnListenerFiltersTimeout": true,
        "trafficDirection": "OUTBOUND"
    }
]
```

**Comment by saitejapakalapati1** [comment]
any workaround for this. not working after renaming port as well

**Comment by howardjohn** [comment]
@saitejapakalapati1 please post a new issue with a config dump (https://github.com/istio/istio/wiki/Troubleshooting-Istio). If you are using Istio 1.8 you are not running into the same issue
